Remove all references to v1.Endpoints from non-network e2e tests

danwinship opened the pull request:
kube-proxy does not look at Endpoints ever, so it is incorrect for a test to assume that there is any correlation between whether Endpoints exist and whether a Service is working. Tests should only be using the v1.Endpoints API if they are explicitly testing the behavior of v1.Endpoints, the Endpoints controller, or the EndpointSlice mirroring controller. There is no reason for any non SIG Network tests to be testing any of those things, so there should be no references to v1.Endpoints in test/e2e outside of test/e2e/network.

Also, simplify some pointlessly complicated e2eservice code.

/kind cleanup
/sig network
/triage accepted
/priority backlog

```release-note
NONE
```

danwinship committed:
Remove all references to v1.Endpoints from non-network e2e tests

kube-proxy does not look at Endpoints ever, so it is incorrect for a
test to assume that there is any correlation between whether Endpoints
exist and whether a Service is working. Tests should only be using the
v1.Endpoints API if they are explicitly testing the behavior of
v1.Endpoints, the Endpoints controller, or the EndpointSlice mirroring
controller. There is no reason for any non SIG Network tests to be
testing any of those things, so there should be no references to
v1.Endpoints in test/e2e outside of test/e2e/network.

Also, simplify some pointlessly complicated e2eservice code.

aojea commented on test/e2e/framework/service/jig.go:
```diff
@@ -355,91 +356,18 @@ func (j *TestJig) WaitForEndpointOnNode(ctx context.Context, nodeName string) er
 
 // waitForAvailableEndpoint waits for at least 1 endpoint to be available till timeout
 func (j *TestJig) waitForAvailableEndpoint(ctx context.Context, timeout time.Duration) error {
-	ctx, cancel := context.WithTimeout(ctx, timeout)
```
this code is correct, a Service MUST have Endponits and EndpointSlices , both, why remove Endpoints from the assertion, this is a Services test

aojea replied:
this is  network e2e test by the way

danwinship replied:
> a Service MUST have Endpoints

Currently. :smiling_imp:

> this is a Services test

It's not a _test_, it's part of the framework. None of the callers of this function care about v1.Endpoints specifically, they care about "the service proxy is aware of the endpoints", and the v1.Endpoints are completely irrelevant to that.

There are other tests in `test/e2e/network` that assert that the Endpoints controller works correctly.

aojea replied:
I feel safer if we keep this double check, things like that help a lot to debug flakes and troubleshoot problems with the endpoints and endpointslice controller because allows you to correlate

danwinship replied:
There are plenty of tests in `tests/e2e/network` that assert that the Endpoints and EndpointSlice controllers give equivalent results. We don't _need_ to do it here as well.

(This PR is the first part of a plan to eventually make the Endpoints controller officially optional (to save disk and CPU), which we can only safely do if we have an e2e job that runs with the Endpoints controller disabled to confirm that everything still works in that configuration, which we can only do if we have a `[Feature:EndpointsController]` tag to wrap around all of the tests that won't work in that job. And we don't want random tests getting skipped in the no-Endpoints-controller case, when they were only trying to figure out "does kube-proxy know about the endpoints of this service?" for which looking at v1.Endpoints isn't even helpful anyway.)

danwinship replied:
https://github.com/kubernetes/enhancements/pull/4975

aojea commented on test/e2e/framework/util.go:
```diff
@@ -684,14 +641,14 @@ func getControlPlaneAddresses(ctx context.Context, c clientset.Interface) ([]str
-			if address.IP != "" {
-				internalIPs = append(internalIPs, address.IP)
+	for _, slice := range slices.Items {
+		for _, endpoint := range slice.Endpoints {
```
ditto, endpoints on endpointslices has the ready condition, to be the same logic as we had with Endpoints we need to check that

aojea commented on test/e2e/network/endpoints.go:
```diff
@@ -14,22 +14,15 @@ See the License for the specific language governing permissions and
 */
 
-package endpoints
+package network
```
lol

danwinship committed:
Remove no-longer-used TestUnderTemporaryNetworkFailure()

This was previously used by some autoscaling tests that have since
been removed. It is somewhat sketchy (and inherently [Disruptive]),
and depends on iptables (so would need to be updated to use nftables
at some point if we were keeping it). Given that it's now unused, just
remove it (as well as some helper functions that are no longer used by
anyone else as well).

ardaguclu commented:
/lgtm
from sig-cli POV

ardaguclu commented:
/approve

aojea commented on test/e2e/framework/endpoints/.import-restrictions:
```diff
@@ -1,12 +0,0 @@
-# This E2E framework sub-package is currently allowed to use arbitrary
-# dependencies except of k/k/pkg, therefore we need to override the
-# restrictions from the parent .import-restrictions file.
```
why do you remove the restrictions here?

aojea replied:
I see, the file is removed

aojea commented:
/lgtm
/approve

Thanks

aojea commented on test/e2e/framework/util.go:
```diff
@@ -416,29 +410,12 @@ func CheckTestingNSDeletedExcept(ctx context.Context, c clientset.Interface, ski
-// WaitForServiceEndpointsNum waits until the amount of endpoints that implement service to expectNum.
-// Some components use EndpointSlices other Endpoints, we must verify that both objects meet the requirements.
+// WaitForServiceEndpointsNum waits until there are EndpointSlices for serviceName
+// containing a total of expectNum endpoints. (If the service is dual-stack, expectNum
+// must count the endpoints of both IP families.)
```
This broke dual stack tests 

> I0121 11:20:30.804569 71111 util.go:431] Unexpected number of Endpoints on Slices, got 4, expected 2
I0121 11:20:31.782539 71111 util.go:418] Waiting for amount of service:node-port-service endpoints to be 2
I0121 11:20:31.792230 71111 util.go:431] Unexpected number of Endpoints on Slices, got 4, expected 2

https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-kind-network-dual/1881658530616840192
